### ISSUE TITLE
[core][rllib] Decreasing CPU utilization with Ray 0.8.4

### ISSUE THREAD

**Issue by annaluo676** [ray-project/ray#7976]
### What is the problem?

*Ray version*: 0.8.4 
*TensorFlow version*: 2.1.0

CPU utilization kept decreasing during an ARS job training.  See below for the metric on a ec2 p2.x16large instance.
<img width="887" alt="Screen Shot 2020-04-10 at 10 04 12 PM" src="https://user-images.githubusercontent.com/45078924/79036017-b919f700-7b78-11ea-99d0-b08859d55a5c.png">

`> ray memory` showed 31 workers pinned in memory, which was expected. There are 124 local references, mostly at `ray/rllib/utils/filter_manager.py:<listcomp>:25`, `ray/rllib/agents/ars/ars.py:<listcomp>:193` and `ray/rllib/agents/ars/ars.py:<listcomp>:314`. Seems that discounted reference counting didn't cause objects to explode.

Besides, Ray dashboard showed almost all `Worker` instead of `Worker.do_rollout()`. 

### Reproduction (REQUIRED)
Script to reproduce:
```
import ray
import os
from ray.rllib.agents.trainer import with_common_config
from ray.rllib.agents.ars.ars import ARSTrainer

gpus = [str(i) for i in list(range(12,16))]
cpus = 32
os.environ["CUDA_VISIBLE_DEVICES"] = ','.join(gpus)

ray.init(
    num_cpus=cpus,
    num_gpus=len(gpus),
    object_store_memory=int(1.5e11))

ars_config = with_common_config({
    "noise_stdev": 0.02,
    "num_rollouts": 50,
    "rollouts_used": 25,
    "num_workers": 31,
    "sgd_stepsize": 0.01,
    "noise_size": 250000000,
    "eval_prob": 0.5,
    "model":{
        "fcnet_hiddens": []  # a linear policy
        }
    })

agent = ARSTrainer(env="CartPole-v0", 
                   config=ars_config)

n_iter = int(1e7)
for i in range(n_iter):    
    result = agent.train()
```

If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://ray.readthedocs.io/en/latest/installation.html).

**Comment by annaluo676** [comment]
0.8.3 also exhibits this lazy worker syndrome; Ray 0.8.2 doesn't.

**Comment by ericl** [comment]
Thanks anna!

**Comment by ericl** [comment]
I can also observe from RLlib's internal metrics, time per iteration gradually increases and CPU load decreases.
![image](https://user-images.githubusercontent.com/14922/79304002-285b5800-7ea5-11ea-9f44-e54fcb7c0b4c.png)

**Comment by ericl** [comment]
`perf top` shows

```
Overhead  Shared Object                                      Symbol
  10.96%  _raylet.so                                         [.] ray::RayObject::IsInPlasmaError
   5.12%  _raylet.so                                         [.] ray::CoreWorkerMemoryStore::GetMemoryStoreStatisticalData
   4.85%  perf                                               [.] __symbols__insert
   3.17%  python3.6                                          [.] _PyEval_EvalFrameDefault
   2.86%  libstdc++.so.6.0.26                                [.] std::string::_M_copy
```

**Comment by ericl** [comment]
This is pointing to a leak in the number of objects in the memory store.

**Comment by ericl** [comment]
@annaluo676 just tested and this patch appears to resolve the slowdown problems.

**Comment by annaluo676** [comment]
Thanks @ericl. According to https://ray.readthedocs.io/en/latest/development.html#compilation I will need to recompile the c++ code changes. `bash build.sh` emitted the following error:
```
INFO: Writing tracer profile to '/home/ubuntu/.cache/bazel/_bazel_ubuntu/4694ec92a350130755ebd31627e14086/command.profile.gz'
ERROR: Skipping '//:ray_pkg': no such package '': error globbing [python/ray/*.py, python/ray/autoscaler/*.py, python/ray/autoscaler/aws/example-full.yaml, python/ray/autoscaler/azure/example-full.yaml, python/ray/autoscaler/gcp/example-full.yaml, python/ray/autoscaler/local/example-full.yaml, python/ray/cloudpickle/*.py, python/ray/core/__init__.py, python/ray/core/generated/__init__.py, python/ray/core/generated/ray/__init__.py, python/ray/core/generated/ray/protocol/__init__.py, python/ray/dashboard/*.py, python/ray/dashboard/metrics_exporter/*.py, python/ray/experimental/*.py, python/ray/util/*.py, python/ray/internal/*.py, python/ray/projects/*.py, python/ray/projects/schema.json, python/ray/projects/templates/cluster_template.yaml, python/ray/projects/templates/project_template.yaml, python/ray/projects/templates/requirements.txt, python/ray/workers/default_worker.py]: /efs/annaluo/ray/python/ray/autoscaler (Too many levels of symbolic links)
WARNING: Target pattern parsing failed.
ERROR: no such package '': error globbing [python/ray/*.py, python/ray/autoscaler/*.py, python/ray/autoscaler/aws/example-full.yaml, python/ray/autoscaler/azure/example-full.yaml, python/ray/autoscaler/gcp/example-full.yaml, python/ray/autoscaler/local/example-full.yaml, python/ray/cloudpickle/*.py, python/ray/core/__init__.py, python/ray/core/generated/__init__.py, python/ray/core/generated/ray/__init__.py, python/ray/core/generated/ray/protocol/__init__.py, python/ray/dashboard/*.py, python/ray/dashboard/metrics_exporter/*.py, python/ray/experimental/*.py, python/ray/util/*.py, python/ray/internal/*.py, python/ray/projects/*.py, python/ray/projects/schema.json, python/ray/projects/templates/cluster_template.yaml, python/ray/projects/templates/project_template.yaml, python/ray/projects/templates/requirements.txt, python/ray/workers/default_worker.py]: /efs/annaluo/ray/python/ray/autoscaler (Too many levels of symbolic links)
```
The error seems to be caused by the symlinks for Rllib. Is this an issue?

**Comment by ericl** [comment]
I think the right way is pip install -e . In the python directory, not
calling that script.

**Comment by annaluo676** [comment]
Unfortunately I was not able to recompile the whole package.  Since the the changes were merged, I can test with the latest wheel. Thanks Eric for looking into this!
